plutus-core/testlib/PlutusCore/Generators/PIR.hs

```diff
@@ -0,0 +1,1752 @@
+runGenTm = runGenTmCustom 0 (error "No custom generator.")
+
+-- | Run a `GenTm` generator with a plug-in custom generator for terms that is included with
+-- the other generators.
+runGenTmCustom :: Int
+               -> (Maybe (Type TyName DefaultUni ())
+                  -> GenTm (Type TyName DefaultUni (), Term TyName Name DefaultUni DefaultFun ()))
+               -> GenTm a
+               -> Gen a
+runGenTmCustom f cg g = sized $ \ n ->
+  runReaderT g $ GenEnv { geSize               = n
+                        , geDatas              = Map.empty
+                        , geTypes              = Map.empty
+                        , geTerms              = Map.empty
+                        , geUnboundUsedTyNames = Set.empty
+                        , geEscaping           = YesEscape
+                        , geCustomGen          = cg
+                        , geCustomFreq         = f
+                        }
+
+-- * Utility functions
+
+-- | Don't allow types to escape from a generator.
+noEscape :: GenTm a -> GenTm a
+noEscape = local $ \env -> env { geEscaping = NoEscape }
+
+-- * Functions for lifting `Gen` stuff to `GenTm`
+
+-- | Lift `Gen` generator to `GenTm` generator. Respects `geSize`.
+liftGen :: Gen a -> GenTm a
+liftGen gen = do
+  sz <- asks geSize
+  lift $ resize sz gen
+
+-- | Lift functor operations like `oneof` from `Gen` to `GenTm`
+liftGenF :: Functor f => (f (Gen a) -> Gen a) -> f (GenTm a) -> GenTm a
+liftGenF oo gs = ReaderT $ \ env -> oo $ fmap (`runReaderT` env) gs
+
+-- | Uniformly choose one of the generators in the list. Requires the
+-- list to be non-empty.
+oneofTm :: [GenTm a] -> GenTm a
+oneofTm = liftGenF oneof
+
+newtype FreqList a = FreqList { unFreqList :: [(Int, a)] }
+  deriving stock Functor
+
+-- | Non-uniformly pick a generator from the list weighted by
+-- the first item in the tuple.
+frequencyTm :: [(Int, GenTm a)] -> GenTm a
+frequencyTm = liftGenF (frequency . unFreqList) . FreqList
+
+-- | Lift a generator from items to lists.
+listTm :: GenTm a -> GenTm [a]
+listTm g = do
+  sz <- asks geSize
+  n  <- liftGen $ choose (0, div sz 3)
+  onSize (`div` n) $ replicateM n g
+
+-- | Generate exactly `n` items of a given generator
+vecTm :: Int -> GenTm a -> GenTm [a]
+vecTm n = sequence . replicate n
+
+-- * Dealing with size
+
+-- | Map a function over the generator size
+onSize :: (Int -> Int) -> GenTm a -> GenTm a
+onSize f = local $ \ env -> env { geSize = f (geSize env) }
+
+-- | Default to the first generator if the size is zero (or negative),
+-- use the second generator otherwise.
+ifSizeZero :: GenTm a -> GenTm a -> GenTm a
+ifSizeZero ifZ nonZ = do
+  n <- asks geSize
+  if n <= 0 then ifZ else nonZ
+
+-- | Locally set the size in a generator
+withSize :: Int -> GenTm a -> GenTm a
+withSize = onSize . const
+
+-- | Split the size between two generators in the ratio specified by
+-- the first two arguments.
+sizeSplit_ :: Int -> Int -> GenTm a -> GenTm b -> (a -> b -> c) -> GenTm c
```

Comment:
same comment about `_`. I won't say use `'` as a mark because I don't really like that either :D naming is hard

Reply:
Why is this CPS-ed instead of just returning a `Gen (a, b)`? And ditto for `sizeSplit`

Reply:
un-CPS-ed it in [this](https://github.com/input-output-hk/plutus/pull/4559/commits/180d8e59ae5b69167c9b1bd911be0403d2e05625) commit. Seems a bit more readable to me. Although I guess I understand the original motivation: `sizeSplit` should be the last thing in a do-block, hence CPS-ing it makes sense. I also tried doing it like this:

```haskell
sizeSplit :: Int -> Int -> GenTm (GenTm a -> GenTm a, GenTm b -> GenTm b)
sizeSplit a b = do
  n <- asks geSize
  let na = (a * n) `div` (a + b)
      nb = (b * n) `div` (a + b)
  (withSize na, withSize nb)
```

example usage:

```haskell
      (resizeL, resizeR) <- sizeSplit 1 7
      tms <- resizeL . mapM genBin $ zip xs as
      (ty, body) <- resizeR . bindTmNames (zip xs as) $ genTerm mty
      let mkBind (x, a, s) tm = TermBind () s
                                  (VarDecl () x a) tm
          b : bs = zipWith mkBind (zip3 xs as ss) tms
      pure (ty, Let () (if r then Rec else NonRec) (b :| bs) body)
```

but that's probably too unsafe, so I `git reset --hard`-ed it.

Reply:
Resolving this conversation, but @MaximilianAlgehed do feel free to revert the commit if you don't like it.